### ISSUE TITLE
readlink: illegal option -- f

### ISSUE THREAD

**Issue by mitchelljfs** [KittyKatt/screenFetch#456]
When I run the program on macOS Sierra 10.12.3 it says the title phrase in normal mode and verbose mode.

Verbose Mode:
```
$ screenfetch -v
:: Finding distro...found as 'Mac OS X '
:: Finding hostname and user...found as 'm@m-macos'
:: Finding kernel version...found as 'x86_64 Darwin 16.4.0'
:: Finding current uptime...found as '3m'
:: Finding current package count...found as '57'
readlink: illegal option -- f
usage: readlink [-n] [file ...]
:: Finding current shell...found as 'bash 3.2.57'
:: Finding current resolution(s)...found as '1280x800'
:: Finding desktop environment...found as 'Aqua'
:: Finding window manager...found as 'Quartz Compositor'
:: Finding window manager theme...found as 'Blue'
:: Finding GTK2 theme...found as 'Not Applicable'
:: Finding GTK3 theme...found as 'Not Applicable'
:: Finding icon theme...found as 'Not Applicable'
:: Finding user font...found as 'SFMono-Regular'
:: Finding current CPU...found as 'Intel Core i5-3210M @ 2.50GHz'
:: Finding current GPU...found as 'Intel HD Graphics 4000 '
:: Finding current RAM usage...found as '2576MiB / 10240MiB'
                               
                 -/+:.          m@m-macos
                :++++.          OS: 64bit Mac OS X 10.12.3 16D32
               /+++/.           Kernel: x86_64 Darwin 16.4.0
       .:-::- .+/:-``.::-       Uptime: 3m
    .:/++++++/::::/++++++/:`    Packages: 57
  .:///////////////////////:`   Shell: bash 3.2.57
  ////////////////////////`     Resolution: 1280x800
 -+++++++++++++++++++++++`      DE: Aqua
 /++++++++++++++++++++++/       WM: Quartz Compositor
 /sssssssssssssssssssssss.      WM Theme: Blue
 :ssssssssssssssssssssssss-     Font: SFMono-Regular
  osssssssssssssssssssssssso/`  CPU: Intel Core i5-3210M @ 2.50GHz
  `syyyyyyyyyyyyyyyyyyyyyyyy+`  GPU: Intel HD Graphics 4000 
   `ossssssssssssssssssssss/    RAM: 2576MiB / 10240MiB
     :ooooooooooooooooooo+.    
      `:+oo+/:-..-:/+o+/-      
```

Normal Mode:

```
$ screenfetch
readlink: illegal option -- f
usage: readlink [-n] [file ...]
                               
                 -/+:.          m@m-macos
                :++++.          OS: 64bit Mac OS X 10.12.3 16D32
               /+++/.           Kernel: x86_64 Darwin 16.4.0
       .:-::- .+/:-``.::-       Uptime: 7m
    .:/++++++/::::/++++++/:`    Packages: 57
  .:///////////////////////:`   Shell: bash 3.2.57
  ////////////////////////`     Resolution: 1280x800
 -+++++++++++++++++++++++`      DE: Aqua
 /++++++++++++++++++++++/       WM: Quartz Compositor
 /sssssssssssssssssssssss.      WM Theme: Blue
 :ssssssssssssssssssssssss-     Font: SFMono-Regular
  osssssssssssssssssssssssso/`  CPU: Intel Core i5-3210M @ 2.50GHz
  `syyyyyyyyyyyyyyyyyyyyyyyy+`  GPU: Intel HD Graphics 4000 
   `ossssssssssssssssssssss/    RAM: 3180MiB / 10240MiB
     :ooooooooooooooooooo+.    
      `:+oo+/:-..-:/+o+/-      `
```

**Comment by StudioEtrange** [comment]
Problem is here, when testing for busybox shell :

```
detectshell () {
	if [[ ! "${shell_type}" ]]; then
		if [[ "${distro}" == "Cygwin" || "${distro}" == "Msys" || "${distro}" == "Haiku" || "${distro}" == "Alpine Linux" || "${OSTYPE}" == "gnu" || "${distro}" == "TinyCore" ]]; then
			shell_type=$(echo "$SHELL" | awk -F'/' '{print $NF}')
		elif readlink -f "$SHELL" | grep -q "busybox"; then
			shell_type="BusyBox"
		else
```

**Comment by darealshinji** [comment]
Does anyone know the right parameters that work with GNU and BSD readlink?

**Comment by tntc** [comment]
I just tested on macOS 10.12.3 and it looks like readlink and stat are the same command, but when called as readlink it appears to only accept -n.

You might be able to cheat it with ls -ls "$SHELL" and look for a link with '->' in it for the symlinked distributions.

**Comment by MacGyver1997** [comment]
I've got the same issue on my MacBook Pro 2015 running with MacOS 10.12.4.
```
***@***-MBP:~ Ω screenfetch -v
:: Finding distro...found as 'Mac OS X '
:: Finding hostname and user...found as '***@***-MBP.***'
:: Finding kernel version...found as 'x86_64 Darwin 16.5.0'
:: Finding current uptime...found as '2d 3h 10m'
:: Finding current package count...found as '31'
readlink: illegal option -- f
usage: readlink [-n] [file ...]
:: Finding current shell...found as 'zsh 5.2'
:: Finding current resolution(s)...found as '2560x1600 1920x1200'
:: Finding desktop environment...found as 'Aqua'
:: Finding window manager...found as 'Quartz Compositor'
:: Finding window manager theme...found as 'Blue'
:: Finding GTK2 theme...found as 'Not Applicable'
:: Finding GTK3 theme...found as 'Not Applicable'
:: Finding icon theme...found as 'Not Applicable'
:: Finding user font...found as 'DroidSansMono'
:: Finding current CPU...found as 'Intel Core i5-5257U @ 2.70GHz'
:: Finding current GPU...found as 'Intel Iris Graphics 6100 '
:: Finding current RAM usage...found as '7832MiB / 16384MiB'
                               
                 -/+:.          ***@***-MBP.***
                :++++.          OS: 64bit Mac OS X 10.12.4 16E195
               /+++/.           Kernel: x86_64 Darwin 16.5.0
       .:-::- .+/:-``.::-       Uptime: 2d 3h 10m
    .:/++++++/::::/++++++/:`    Packages: 31
  .:///////////////////////:`   Shell: zsh 5.2
  ////////////////////////`     Resolution: 2560x1600 1920x1200
 -+++++++++++++++++++++++`      DE: Aqua
 /++++++++++++++++++++++/       WM: Quartz Compositor
 /sssssssssssssssssssssss.      WM Theme: Blue
 :ssssssssssssssssssssssss-     Font: DroidSansMono
  osssssssssssssssssssssssso/`  CPU: Intel Core i5-5257U @ 2.70GHz
  `syyyyyyyyyyyyyyyyyyyyyyyy+`  GPU: Intel Iris Graphics 6100 
   `ossssssssssssssssssssss/    RAM: 7832MiB / 16384MiB
     :ooooooooooooooooooo+.    
      `:+oo+/:-..-:/+o+/-      
                               
***@***-MBP:~ Ω 

```

I heard something about "greadlink". I installed it, but i'm not able to get it running.

**Comment by tntc** [comment]
I was just tinkering in Debian and ran screenfetch inside of ash (busybox's shell): $SHELL was still /bin/bash, but it correctly stated busybox as the shell type because of logic in the else section (line 1496).  It looks like it was introduced in [this commit](https://github.com/KittyKatt/screenFetch/commit/fa79fa91feed3d5e89e67af99d329a040d376281) which removed logic for detecting TinyCore Linux.  Perhaps that section could go in the further OS detection section? I'll try to kick together a patch to show what I mean.

**Comment by tntc** [comment]
Here's the patch I cobbled together.

`Index: screenfetch-dev
===================================================================
--- screenfetch-dev	(revision 1edfc3aee9681d9bef0732f55aec36980b09da1c)
+++ screenfetch-dev	(revision )
@@ -1491,13 +1491,13 @@
 	if [[ ! "${shell_type}" ]]; then
 		if [[ "${distro}" == "Cygwin" || "${distro}" == "Msys" || "${distro}" == "Haiku" || "${distro}" == "Alpine Linux" || "${OSTYPE}" == "gnu" || "${distro}" == "TinyCore" || "${distro}" == "Raspbian" ]]; then
 			shell_type=$(echo "$SHELL" | awk -F'/' '{print $NF}')
-		elif readlink -f "$SHELL" | grep -q "busybox"; then
-			shell_type="BusyBox"
 		else
 			if [[ "${OSTYPE}" == "linux-gnu" || "${OSTYPE}" == "linux" || "${OSTYPE}" == "linux-musl" ]]; then
 				shell_type=$(cat /proc/$PPID/cmdline|tr '\0' '\n'|head -n 1)
 			elif [[ "${distro}" == "Mac OS X" || "${distro}" == "DragonFlyBSD" || "${distro}" == "FreeBSD" || "${distro}" == "OpenBSD" || "${distro}" == "NetBSD" ]]; then
 				shell_type=$(ps -p $PPID -o command | tail -1)
+			elif readlink -f "$SHELL" | grep -q "busybox"; then
+			    shell_type="BusyBox"
 			else
 				shell_type=$(ps -p $(ps -p $PPID | awk '$1 !~ /PID/ {print $1}') | awk 'FNR>1 {print $1}')
 			fi
`

Does anyone know what prompted the initial change? What OS besides TinyCore was using busybox and failing to detect it? That might let me hack together a more robust solution.

**Comment by mymtwcom** [comment]
You can suppress errors with `-E` option. So just run this e.g `screenfetch -E -A "Arch Linux"`

**Comment by darealshinji** [comment]
Is it now fixed?

**Comment by expenses** [comment]
@darealshinji Yep, the version of screenfetch in the repo works just fine. Homebrew hasn't updated yet though.

**Comment by CyberManifest** [comment]
Homebrew still hasn't updated ...
http://dpaste.com/1WZVM0F

**Comment by MacGyver1997** [comment]
For me too. I still get the „readlink: illegal option“
I actually wrote an alias just to run screenfetch without this error.
So I start up the Terminal with this alias and I‘m fine right now. But it‘s only a workaround.

**Comment by CyberManifest** [comment]
Yeah, you mean with the "-E" switch? Yeah, that's a good idea; but as you mentioned it's "only a workaround." I just wanted to make it aware that after some extensive time it's still an issue. I'll probably use the same workaround as you mention, just wish they would update Homebrew with the corrected version.

**Comment by MacGyver1997** [comment]
Hi again,
I guess I solved the problem on my side. Actually I think it's still a little workaround, but it should be better than ignoring the errors with "-E".

I installed "greadlink" via homebrew (the GNU Version of readlink) and made a symbolic link from greadlink to readlink.

In the homebrew directory `/usr/local/bin`, I did `ln -s greadlink readlink` to create the link. At the next start of the shell the error does not appear anymore.
Just for a test I did `which readlink` and got this as return: `/usr/local/bin/readlink`.

This resolved the problem for me.
